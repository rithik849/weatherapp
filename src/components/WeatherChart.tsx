import React from "react"
import * as d3 from "d3"
import { DateTime } from "luxon"


interface GraphProps<T extends readonly string[] = readonly string[]> {

    time : T,
    temperature : {[K in keyof T] : number},
    x_label : string,
    y_label : string,
    x_units ?: string,
    y_units ?: string
    

}

type GraphPoint = [Date,number]

interface GraphState {
    points : GraphPoint[]
    x_label : string,
    y_label : string,
    x_units ?: string,
    y_units ?: string
}

class WeatherChart extends React.Component<GraphProps,GraphState> {
    constructor(props : GraphProps){
        super(props)
        this.state = {
            points : props.time.map((v,i) => [DateTime.fromFormat(v,"T").toJSDate() ,props.temperature[i]]),
            x_label : props.x_label,
            y_label : props.y_label,
            x_units : props.x_units,
            y_units : props.y_units

        }

    }

    drawChart(){
        if (this.state.points.length === 0){
            throw new Error('No data to plot!')
        }

        const margin = {top: 10, right: 30, bottom: 50, left: 60}
        const dimensions = {width : 1080 - margin.left - margin.right , height : 300 - margin.top - margin.bottom}

        const svg = d3.select(".graph")
        .style("position","relative")
        .append("svg")
        .style("min-width", dimensions.width + margin.left + margin.right + "px")
        .style("min-height",dimensions.height + margin.top + margin.bottom + "px")
        .style("background-color","red")
        .style("position","relative")
        .attr("id","plot")
        .append("g")

        const minmaxDate : [Date,Date]= d3.extent(this.state.points, (d : GraphPoint) => d[0]) as [Date,Date]

        // Ensure 0 is lowest temperature displayed if all temperatures are 0 or greater
        const minmaxTemperature : [number,number] = d3.extent(this.state.points, (d : GraphPoint) => d[1]) as [number,number]
        if (minmaxTemperature[0] > 0){
            minmaxTemperature[0] = 0
        }

        const xAxis : d3.ScaleTime<number,number,never> = d3.scaleTime().domain(minmaxDate).rangeRound([0,dimensions.width])

        const yAxis : d3.ScaleLinear<number,number> = d3.scaleLinear().domain(minmaxTemperature).rangeRound([dimensions.height,0])

        // Plot x axis and label
        svg.append("g")
        .attr("color","black")
        .attr("transform","translate("+margin.left+","+(margin.top+dimensions.height)+")")
        .call(
            d3.axisBottom(xAxis).tickFormat(
                (d) => d3.timeFormat('%H:%M')(d as Date)
                ).ticks(24)
            )

        svg.append("text")
        .attr("class","x-label")
        .attr("x",margin.left + dimensions.width/2)
        .attr("y",(margin.top+dimensions.height + 40) + "px")
        .style("text-anchor","middle")
        .text(this.state.x_label)
        

        // Plot y axis and label
        svg.append("g")
        .attr("color","black")
        .attr("transform","translate("+margin.left+","+margin.top+")")
        .call(
            d3.axisLeft(yAxis)
            )

        svg.append("text")
        .attr("class", "y-label")
        .attr("transform", "rotate(-90)")
        .attr("x",-(margin.top + dimensions.height/2))
        .attr("y",(margin.left - 40) + "px")
        .style("text-anchor", "middle")
        .text(this.state.y_label)




        // Tooltip
        // create a tooltip
        
        
        const Tooltip = d3.select(".graph")
        .append("div")
        .attr("class", "tooltip")
        .style("visibility", "hidden")
        .style("background-color", "blue")
        .style("border", "solid")
        .style("border-width", "2px")
        .style("border-radius", "5px")
        .style("padding", "5px")
        .style("position","absolute")

        const x_units : string | undefined = this.state.x_units
        const y_units : string | undefined = this.state.y_units
            

        // Three function that change the tooltip when user hover / move / leave a cell
        const mouseover = function(this : SVGElement ,event : any, d : GraphPoint) {
            Tooltip.style("visibility", "visible").style("opacity",0.5)
        }

        const mousemove = function(this : SVGElement ,event : any, d : GraphPoint) {
            
            const display : string = d3.timeFormat("%H:%M")(d[0] as Date) + x_units +  "," + d[1] + y_units 
            Tooltip
            .html(display)
            .style("left", (event.pageX) + "px")
            .style("top", (event.pageY) + "px")
        }

        const mouseleave = function(this : SVGElement ,event : any, d : GraphPoint) {
            Tooltip
            .style("visibility", "hidden")
        }

                // Plot graph

        svg.append("path")
        .datum(this.state.points)
        .attr("id","line")
        .attr("fill", "none")
        .attr("transform","translate("+margin.left+","+margin.top+")")
        .attr("stroke", "black")
        .attr("stroke-width", 1.5)
        .attr("d", d3.line(
            d => {return xAxis(d[0])}, d => {return yAxis(d[1])}
            )
        )

        svg.append('g')
        .selectAll("dot").data(this.state.points ).enter()
        .append("circle")
          .attr("cx", function (d : GraphPoint) { return xAxis(d[0]); } )
          .attr("cy", function (d : GraphPoint) { return yAxis(d[1]); } )
          .attr("r", 7)
          .style("fill", "#69b3a2")
          .style("opacity", 1)
          .style("stroke", "white")
          .attr("transform", "translate(" + margin.left + "," + margin.top + ")")
        .on("mouseover",mouseover)
        .on("mousemove",mousemove)
        .on("mouseleave",mouseleave)
    
    }

    componentWillUnmount(): void {
        d3.selectAll('svg').remove()
    }



    componentDidMount(): void {
        try{
            this.drawChart()
        }
        catch(e){
            console.error(e)
        }
        

    }

    render(): React.ReactNode {
        return <div className = 'graph'></div>
    }

}


export default WeatherChart