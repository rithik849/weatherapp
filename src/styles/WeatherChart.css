
/* * {
    outline: 1px solid blue;
} */

#root{
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    overflow-x : hidden;
    height : 100%;
}

@media screen and (width < 1080px){
    .graph{
        display: flex;
        background-color: gray;
        width : 100dvw;
        height : auto;
        overflow-x: scroll;
        /* min-width : 1080px; */
    }

    body{
        overflow-x: hidden;

    }

}



