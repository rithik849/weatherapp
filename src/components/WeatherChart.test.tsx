import { render, waitFor, screen, fireEvent, RenderResult, cleanup } from "@testing-library/react"
import WeatherChart from "./WeatherChart"
import { DateTime } from "luxon"
import { UserEvent } from '@testing-library/user-event/dist/types/setup/setup'
import userEvent from "@testing-library/user-event"
import { min } from "d3"

let renderResult : RenderResult | null = null;

const dummy_time_data : readonly string[] = [
  "00:00",
  "01:00",
  "02:00",
  "03:00",
  "04:00",
  "05:00",
  "06:00",
  "07:00",
  "08:00",
  "09:00",
  "10:00",
  "11:00",
  "12:00",
  "13:00",
  "14:00",
  "15:00",
  "16:00",
  "17:00",
  "18:00",
  "19:00",
  "20:00",
  "21:00",
  "22:00",
  "23:00"
]

const dummy_temperature_data : number[] = [2,4,6,8,10,12,14,16,18,20,22,24,26,28,30,32,34,36,38,40,42,44,46,48]

describe("Test Weather Chart Plot", () => {
  
    test("Graph tooltip renders correctly", async () => {
      const {asFragment,getAllByRole, getByText, getByTestId} = render(<WeatherChart time = {dummy_time_data} temperature = {dummy_temperature_data} x_label ="Time" y_label ="Temperature" y_units="C"/>)
      expect(asFragment()).toMatchSnapshot()
      let user : UserEvent = userEvent.setup()

      for (let i = 0; i < dummy_temperature_data.length; i++){
        let element : HTMLElement = getAllByRole("application",{hidden:true, name: "point"})[i]

        await userEvent.hover(element)
        element = getByText(/[0-9][0-9]:[0-9][0-9],([0-9]*)/)
        expect(element.innerHTML).toMatch(/[0-9][0-9]:[0-9][0-9],([0-9]*)/)
        expect(element.innerHTML).toStrictEqual(dummy_time_data[i] + "," + dummy_temperature_data[i]+"C")
      }
    }
    
    )

    test("Correct X and Y labels",async () => {
        const {getByText} = render(<WeatherChart time = {dummy_time_data} temperature = {dummy_temperature_data} x_label ="Time" y_label ="Temperature" y_units="C"/>)
        let y_label =  getByText("Temperature (C)")
        let x_label = getByText("Time")
        

        expect(x_label).toBeDefined()
        expect(x_label.innerHTML).toStrictEqual("Time")

        expect(y_label).toBeDefined()
        expect(y_label.innerHTML).toStrictEqual("Temperature (C)")
          
    })

    test("Correct Scale of y Axis at 0",async () => {
      let {queryByText} = render(<WeatherChart time = {dummy_time_data} temperature = {dummy_temperature_data} x_label ="Time" y_label ="Temperature" y_units="C"/>)
      let min_y =  queryByText("0")

      expect(min_y).not.toBeNull()
    })
    
    test("Correct Scale of y Axis less than 0", async() => {
      let {debug,getByText,findByText} = render(<WeatherChart time = {dummy_time_data} temperature = {dummy_temperature_data.map(d => d-24)} x_label ="Time" y_label ="Temperature" y_units="C"/>)
      // debug()
      // Need to account for d3 minus locale https://github.com/d3/d3-format/issues/62
      let min_y = getByText(/(-|âˆ’)22/) 

      expect(min_y).not.toBeNull()

      let max_y = getByText(/24/)
      expect(max_y).not.toBeNull()
    })

    afterEach(()=>{
      cleanup()
    })

})


export {dummy_time_data, dummy_temperature_data}