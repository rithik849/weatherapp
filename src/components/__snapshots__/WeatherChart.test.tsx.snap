// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Weather Chart Plot Graph tooltip renders correctly 1`] = `
<DocumentFragment>
  <div
    class="graph"
  >
    <svg
      id="plot"
      style="min-width: 1080px; min-height: 600px; background-color: red;"
    >
      <g>
        <g
          color="black"
          fill="none"
          font-family="sans-serif"
          font-size="10"
          style="font-size: 0.8rem;"
          text-anchor="middle"
          transform="translate(60,540)"
        >
          <path
            class="domain"
            d="M0.5,6V0.5H940.5V6"
            stroke="currentColor"
          />
          <g
            class="tick"
            opacity="1"
            transform="translate(0.5,0)"
          >
            <line
              stroke="currentColor"
              y2="6"
            />
            <text
              dy="0.71em"
              fill="currentColor"
              y="9"
            >
              00:00
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(41.5,0)"
          >
            <line
              stroke="currentColor"
              y2="6"
            />
            <text
              dy="0.71em"
              fill="currentColor"
              y="9"
            >
              01:00
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(82.5,0)"
          >
            <line
              stroke="currentColor"
              y2="6"
            />
            <text
              dy="0.71em"
              fill="currentColor"
              y="9"
            >
              02:00
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(123.5,0)"
          >
            <line
              stroke="currentColor"
              y2="6"
            />
            <text
              dy="0.71em"
              fill="currentColor"
              y="9"
            >
              03:00
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(163.5,0)"
          >
            <line
              stroke="currentColor"
              y2="6"
            />
            <text
              dy="0.71em"
              fill="currentColor"
              y="9"
            >
              04:00
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(204.5,0)"
          >
            <line
              stroke="currentColor"
              y2="6"
            />
            <text
              dy="0.71em"
              fill="currentColor"
              y="9"
            >
              05:00
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(245.5,0)"
          >
            <line
              stroke="currentColor"
              y2="6"
            />
            <text
              dy="0.71em"
              fill="currentColor"
              y="9"
            >
              06:00
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(286.5,0)"
          >
            <line
              stroke="currentColor"
              y2="6"
            />
            <text
              dy="0.71em"
              fill="currentColor"
              y="9"
            >
              07:00
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(327.5,0)"
          >
            <line
              stroke="currentColor"
              y2="6"
            />
            <text
              dy="0.71em"
              fill="currentColor"
              y="9"
            >
              08:00
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(368.5,0)"
          >
            <line
              stroke="currentColor"
              y2="6"
            />
            <text
              dy="0.71em"
              fill="currentColor"
              y="9"
            >
              09:00
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(409.5,0)"
          >
            <line
              stroke="currentColor"
              y2="6"
            />
            <text
              dy="0.71em"
              fill="currentColor"
              y="9"
            >
              10:00
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(450.5,0)"
          >
            <line
              stroke="currentColor"
              y2="6"
            />
            <text
              dy="0.71em"
              fill="currentColor"
              y="9"
            >
              11:00
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(490.5,0)"
          >
            <line
              stroke="currentColor"
              y2="6"
            />
            <text
              dy="0.71em"
              fill="currentColor"
              y="9"
            >
              12:00
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(531.5,0)"
          >
            <line
              stroke="currentColor"
              y2="6"
            />
            <text
              dy="0.71em"
              fill="currentColor"
              y="9"
            >
              13:00
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(572.5,0)"
          >
            <line
              stroke="currentColor"
              y2="6"
            />
            <text
              dy="0.71em"
              fill="currentColor"
              y="9"
            >
              14:00
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(613.5,0)"
          >
            <line
              stroke="currentColor"
              y2="6"
            />
            <text
              dy="0.71em"
              fill="currentColor"
              y="9"
            >
              15:00
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(654.5,0)"
          >
            <line
              stroke="currentColor"
              y2="6"
            />
            <text
              dy="0.71em"
              fill="currentColor"
              y="9"
            >
              16:00
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(695.5,0)"
          >
            <line
              stroke="currentColor"
              y2="6"
            />
            <text
              dy="0.71em"
              fill="currentColor"
              y="9"
            >
              17:00
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(736.5,0)"
          >
            <line
              stroke="currentColor"
              y2="6"
            />
            <text
              dy="0.71em"
              fill="currentColor"
              y="9"
            >
              18:00
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(777.5,0)"
          >
            <line
              stroke="currentColor"
              y2="6"
            />
            <text
              dy="0.71em"
              fill="currentColor"
              y="9"
            >
              19:00
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(817.5,0)"
          >
            <line
              stroke="currentColor"
              y2="6"
            />
            <text
              dy="0.71em"
              fill="currentColor"
              y="9"
            >
              20:00
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(858.5,0)"
          >
            <line
              stroke="currentColor"
              y2="6"
            />
            <text
              dy="0.71em"
              fill="currentColor"
              y="9"
            >
              21:00
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(899.5,0)"
          >
            <line
              stroke="currentColor"
              y2="6"
            />
            <text
              dy="0.71em"
              fill="currentColor"
              y="9"
            >
              22:00
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(940.5,0)"
          >
            <line
              stroke="currentColor"
              y2="6"
            />
            <text
              dy="0.71em"
              fill="currentColor"
              y="9"
            >
              23:00
            </text>
          </g>
        </g>
        <text
          class="x-label"
          style="text-anchor: middle;"
          x="530"
          y="580px"
        >
          Time
        </text>
        <g
          color="black"
          fill="none"
          font-family="sans-serif"
          font-size="10"
          style="font-size: 1rem;"
          text-anchor="end"
          transform="translate(60,60)"
        >
          <path
            class="domain"
            d="M-6,480.5H0.5V0.5H-6"
            stroke="currentColor"
          />
          <g
            class="tick"
            opacity="1"
            transform="translate(0,480.5)"
          >
            <line
              stroke="currentColor"
              x2="-6"
            />
            <text
              dy="0.32em"
              fill="currentColor"
              x="-9"
            >
              0
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(0,430.5)"
          >
            <line
              stroke="currentColor"
              x2="-6"
            />
            <text
              dy="0.32em"
              fill="currentColor"
              x="-9"
            >
              5
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(0,380.5)"
          >
            <line
              stroke="currentColor"
              x2="-6"
            />
            <text
              dy="0.32em"
              fill="currentColor"
              x="-9"
            >
              10
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(0,330.5)"
          >
            <line
              stroke="currentColor"
              x2="-6"
            />
            <text
              dy="0.32em"
              fill="currentColor"
              x="-9"
            >
              15
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(0,280.5)"
          >
            <line
              stroke="currentColor"
              x2="-6"
            />
            <text
              dy="0.32em"
              fill="currentColor"
              x="-9"
            >
              20
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(0,230.5)"
          >
            <line
              stroke="currentColor"
              x2="-6"
            />
            <text
              dy="0.32em"
              fill="currentColor"
              x="-9"
            >
              25
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(0,180.5)"
          >
            <line
              stroke="currentColor"
              x2="-6"
            />
            <text
              dy="0.32em"
              fill="currentColor"
              x="-9"
            >
              30
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(0,130.5)"
          >
            <line
              stroke="currentColor"
              x2="-6"
            />
            <text
              dy="0.32em"
              fill="currentColor"
              x="-9"
            >
              35
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(0,80.5)"
          >
            <line
              stroke="currentColor"
              x2="-6"
            />
            <text
              dy="0.32em"
              fill="currentColor"
              x="-9"
            >
              40
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(0,30.5)"
          >
            <line
              stroke="currentColor"
              x2="-6"
            />
            <text
              dy="0.32em"
              fill="currentColor"
              x="-9"
            >
              45
            </text>
          </g>
          <g
            class="tick"
            opacity="1"
            transform="translate(0,0.5)"
          >
            <line
              stroke="currentColor"
              x2="-6"
            />
            <text
              dy="0.32em"
              fill="currentColor"
              x="-9"
            >
              48
            </text>
          </g>
        </g>
        <text
          class="y-label"
          style="text-anchor: middle;"
          transform="rotate(-90)"
          x="-300"
          y="20px"
        >
          Temperature (C)
        </text>
        <path
          d="M0,460L41,440L82,420L123,400L163,380L204,360L245,340L286,320L327,300L368,280L409,260L450,240L490,220L531,200L572,180L613,160L654,140L695,120L736,100L777,80L817,60L858,40L899,20L940,0"
          fill="none"
          id="line"
          stroke="black"
          stroke-width="1.5"
          transform="translate(60,60)"
        />
        <g>
          <circle
            aria-label="point"
            cx="0"
            cy="460"
            r="7"
            role="application"
            style="fill: #69b3a2; opacity: 1; stroke: white;"
            transform="translate(60,60)"
          />
          <circle
            aria-label="point"
            cx="41"
            cy="440"
            r="7"
            role="application"
            style="fill: #69b3a2; opacity: 1; stroke: white;"
            transform="translate(60,60)"
          />
          <circle
            aria-label="point"
            cx="82"
            cy="420"
            r="7"
            role="application"
            style="fill: #69b3a2; opacity: 1; stroke: white;"
            transform="translate(60,60)"
          />
          <circle
            aria-label="point"
            cx="123"
            cy="400"
            r="7"
            role="application"
            style="fill: #69b3a2; opacity: 1; stroke: white;"
            transform="translate(60,60)"
          />
          <circle
            aria-label="point"
            cx="163"
            cy="380"
            r="7"
            role="application"
            style="fill: #69b3a2; opacity: 1; stroke: white;"
            transform="translate(60,60)"
          />
          <circle
            aria-label="point"
            cx="204"
            cy="360"
            r="7"
            role="application"
            style="fill: #69b3a2; opacity: 1; stroke: white;"
            transform="translate(60,60)"
          />
          <circle
            aria-label="point"
            cx="245"
            cy="340"
            r="7"
            role="application"
            style="fill: #69b3a2; opacity: 1; stroke: white;"
            transform="translate(60,60)"
          />
          <circle
            aria-label="point"
            cx="286"
            cy="320"
            r="7"
            role="application"
            style="fill: #69b3a2; opacity: 1; stroke: white;"
            transform="translate(60,60)"
          />
          <circle
            aria-label="point"
            cx="327"
            cy="300"
            r="7"
            role="application"
            style="fill: #69b3a2; opacity: 1; stroke: white;"
            transform="translate(60,60)"
          />
          <circle
            aria-label="point"
            cx="368"
            cy="280"
            r="7"
            role="application"
            style="fill: #69b3a2; opacity: 1; stroke: white;"
            transform="translate(60,60)"
          />
          <circle
            aria-label="point"
            cx="409"
            cy="260"
            r="7"
            role="application"
            style="fill: #69b3a2; opacity: 1; stroke: white;"
            transform="translate(60,60)"
          />
          <circle
            aria-label="point"
            cx="450"
            cy="240"
            r="7"
            role="application"
            style="fill: #69b3a2; opacity: 1; stroke: white;"
            transform="translate(60,60)"
          />
          <circle
            aria-label="point"
            cx="490"
            cy="220"
            r="7"
            role="application"
            style="fill: #69b3a2; opacity: 1; stroke: white;"
            transform="translate(60,60)"
          />
          <circle
            aria-label="point"
            cx="531"
            cy="200"
            r="7"
            role="application"
            style="fill: #69b3a2; opacity: 1; stroke: white;"
            transform="translate(60,60)"
          />
          <circle
            aria-label="point"
            cx="572"
            cy="180"
            r="7"
            role="application"
            style="fill: #69b3a2; opacity: 1; stroke: white;"
            transform="translate(60,60)"
          />
          <circle
            aria-label="point"
            cx="613"
            cy="160"
            r="7"
            role="application"
            style="fill: #69b3a2; opacity: 1; stroke: white;"
            transform="translate(60,60)"
          />
          <circle
            aria-label="point"
            cx="654"
            cy="140"
            r="7"
            role="application"
            style="fill: #69b3a2; opacity: 1; stroke: white;"
            transform="translate(60,60)"
          />
          <circle
            aria-label="point"
            cx="695"
            cy="120"
            r="7"
            role="application"
            style="fill: #69b3a2; opacity: 1; stroke: white;"
            transform="translate(60,60)"
          />
          <circle
            aria-label="point"
            cx="736"
            cy="100"
            r="7"
            role="application"
            style="fill: #69b3a2; opacity: 1; stroke: white;"
            transform="translate(60,60)"
          />
          <circle
            aria-label="point"
            cx="777"
            cy="80"
            r="7"
            role="application"
            style="fill: #69b3a2; opacity: 1; stroke: white;"
            transform="translate(60,60)"
          />
          <circle
            aria-label="point"
            cx="817"
            cy="60"
            r="7"
            role="application"
            style="fill: #69b3a2; opacity: 1; stroke: white;"
            transform="translate(60,60)"
          />
          <circle
            aria-label="point"
            cx="858"
            cy="40"
            r="7"
            role="application"
            style="fill: #69b3a2; opacity: 1; stroke: white;"
            transform="translate(60,60)"
          />
          <circle
            aria-label="point"
            cx="899"
            cy="20"
            r="7"
            role="application"
            style="fill: #69b3a2; opacity: 1; stroke: white;"
            transform="translate(60,60)"
          />
          <circle
            aria-label="point"
            cx="940"
            cy="0"
            r="7"
            role="application"
            style="fill: #69b3a2; opacity: 1; stroke: white;"
            transform="translate(60,60)"
          />
        </g>
      </g>
    </svg>
    <div
      class="tooltip"
      data-testid="tooltip"
      name="tooltip"
      style="visibility: hidden; background-color: blue; border: solid; border-width: 2px; border-radius: 5px; padding: 5px; position: absolute;"
    />
  </div>
</DocumentFragment>
`;
